src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,359 @@
+        // or displayableId, since we didn't get idtoken back, we don't store user as part of key in this
+        // case. Try again to find token with no user.
+        if (mMrrtTokenCacheItem == null) {
+            if (lookUpCachedTokenWithNoUserForADFS()) {
+                mMrrtTokenCacheItem = mTokenCacheAccessor.getMRRTItem(mAuthRequest.getClientId(), null);
+                // ADFS doesn't support Family of Client Id, if we still cannot find anything, stop trying. 
+                if (mMrrtTokenCacheItem == null) {
+                    return null;
+                }
+            } else {
+                Logger.v(TAG, "MRRT token does not exist, try with FRT");
+                return tryFRT(AuthenticationConstants.MS_FAMILY_ID, null);
+            }
+        } 
+        
+        if (!StringExtensions.IsNullOrBlank(mMrrtTokenCacheItem.getFamilyClientId()) && !mAttemptedWithFRT) {
+            Logger.v(TAG, "MRRT item exists but it's also a FRT, try with FRT.");
+            return tryFRT(mMrrtTokenCacheItem.getFamilyClientId(), null);
+        }
+        
+        Logger.v(TAG, "Send request to use MRRT for new AT.");
+        AuthenticationResult mrrtResult = acquireTokenWithCachedItem(mMrrtTokenCacheItem);
+        if (containOauthError(mrrtResult)) {
+            // If MRRT fails, we still want to retry on FRT in case there is one there. 
+            // MRRT may not be marked as FRT, hard-code it as "1" in this case. 
+            final String familyClientId = StringExtensions.IsNullOrBlank(mMrrtTokenCacheItem.getFamilyClientId()) ? 
+                    AuthenticationConstants.MS_FAMILY_ID : mMrrtTokenCacheItem.getFamilyClientId();
+
+            // Rest mMrrtTokenCacheItem to avoid the fallback when FRT request fails.
+            mMrrtTokenCacheItem = null;
+            mrrtResult = tryFRT(familyClientId, mrrtResult);
+        }
+        
+        return mrrtResult;
+    }
+    
+    /**
+     * Attempt to get access token with FRT. 
+     */
+    private AuthenticationResult tryFRT(final String familyClientId, final AuthenticationResult mrrtResult) 
+            throws AuthenticationException {
+        if (mAttemptedWithFRT) {
+            return mrrtResult;
+        }
+        
+        mAttemptedWithFRT = true;
+        final TokenCacheItem frtTokenCacheItem = mTokenCacheAccessor.getFRTItem(familyClientId, getUser());
+        
+        if (frtTokenCacheItem  == null) {
+            if (mMrrtTokenCacheItem != null) {
+                // if mrrt token cache item is already used for token request, it will be reset as null
+                Logger.v(TAG, "FRT cache item does not exist, fall back to try MRRT.");
+                return tryMRRT();
+            } else {
+                // No FRT existed, and we've already tried to get new AT with MRRT, returning MRRT token request result in this case. 
+                return mrrtResult;
```

Comment:
should be mrrtFailureResult
consider having a new method useMRRT handle the case.